docs/RFCS/20200624_cert_revocation.md

```diff
@@ -0,0 +1,144 @@
+- Feature Name: TLS client certificate revocation
+- Status: draft
+- Start Date: 2020-06-24
+- Authors: knz
+- RFC PR: [#50602](https://github.com/cockroachdb/cockroach/pull/50602)
+- Cockroach Issue: [#29641](https://github.com/cockroachdb/cockroach/issues/29641)
+
+# Summary
+
+This RFC proposes to introduce a certificate revocation list (CRL) in
+CockroachDB to validate against TLS client certificates used to
+authenticate SQL client connections.
+
+The motivation is to enable revoking certificates earlier than their
+expiration date, in case a certificate was compromised or the user
+account is temporarily blocked from access (e.g. account suspension).
+
+This is also required for various compliance checklists.
+
+The technical proposal is twofold:
+
+- introduce a new cluster setting containing
+  the revocation list: `server.user_login.certificate_revocation_list`.
+  This is the main CRL store used during authentication.
+
+- introduce two additional cluster setting to refresh
+  the primary setting above periodically:
+  `server.user_login.certificate_revocation_list.update_url`,
+  `server.user_login.certificate_revocation_list.update_interval`.
+
+# Motivation
+
+see above
+
+# Guide-level explanation
+
+After a cluster has been set up, a security event may occur that
+requires the operator to revoke a client TLS certificate. This event
+could be a compromise (i.e. certificate lands in the wrong hands) or
+the deployment of corporate IT / compliance processes that require
+CRLs in all networked services.
+
+Technically a CRL is a file containing one or more *revocation
+certificates* encoded using a standard format (x509 of some sorts).
+
+If the CRL is provided externally as a collection of discrete files,
+they can be combined together into a single file via an `openssl` command.
+
+Once a single-file CRL must be loaded into CockroachDB, the operator can do either of the following:
+
+- they can set the cluster setting
+  `server.user_login.certificate_revocation_list` manually to the
+  contents of their CRL.
+
+  When updating the list manually, the operator should ensure they
+  have a process to update this list whenever their CRL changes.
+
+- or, they can set up a web server in their infrastructure that
+  is reachable from CockroachDB nodes, and have that server serve
+  the CRL at a particular URL. Then, they can set the cluster setting
+  `server.user_login.certificate_revocation_list.update_url` to
+  the location of the CRL.
+
+  The companion setting
+  `server.user_login.certificate_revocation_list.update_interval`
+  determines how often CockroachDB fetches a new CRL from the
+  URL. This setting is set to **1 hour** by default.
+
+
+Note:
+
+- if the operator sets the `server.user_login.certificate_revocation_list` manually,
+  and then configures `...update_url`, their manual configuration will be overwritten
```

Comment:
This seems like a footgun, no? If I specify it manually, it should stick forever (i.e. the data sources should merge).

Note that there isn't an issue if we say that you only get to provide a URL, but we have a custom scheme such as inline://revoke1,revoke2,...` (or even say "if it doesn't look like a URL but looks like a list, it's a list")?

Reply:
The other option would be to produce an error to the operator if they try to set both and only data source to be active at a time. This might be less surprising in practice to the operator and prevent errors.